### PR TITLE
Add metric for timestamp of the files file_sd is using.

### PR DESCRIPTION
**Pull request by KalivarapuReshma**
@Gouthamve  @brian-brazil 

This is for issue [#2757](https://github.com/prometheus/prometheus/issues/2757)

### CODE REVIEW

**Comment by brian-brazil** [comment]
What happens if a file disappears?

File: discovery/file/file.go [hunk 1]
```diff
@@ -33,6 +34,48 @@ import (
 
 const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
 
+//TimeStampManager is a Custom Collector for Timestamps of the files
+type TimeStampManager struct {
```
Comment (Gouthamve) [inline discovery/file/file.go]: Name it TimestampCollector?

File: discovery/file/file.go [hunk 2]
```diff
@@ -33,6 +34,48 @@ import (
...
+	Description *prometheus.Desc
+}
+
+//Describe method sends the description to the channel
+func (t *TimeStampManager) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+//GetFiles changes the filenames of the struct to the paths returned by listfiles()
+func (t *TimeStampManager) GetFiles(files []string) {
```
Comment (Gouthamve) [inline discovery/file/file.go]: SetFiles instead of GetFiles?

File: discovery/file/file.go [hunk 3]
```diff
@@ -33,6 +34,48 @@ import (
...
+	t.filenames = files
+}
+
+//Collect creates constant metrics for each file with last modified time of the file
+func (t *TimeStampManager) Collect(ch chan<- prometheus.Metric) {
+	files := t.filenames
+	for i := 0; i < len(files); i++ {
+		info, _ := os.Stat(files[i])
```
Comment (Gouthamve) [inline discovery/file/file.go]: Always handle errors.

File: discovery/file/file.go [hunk 4]
```diff
@@ -33,6 +34,48 @@ import (
...
+		ch <- prometheus.MustNewConstMetric(
+			t.Description,
+			prometheus.CounterValue,
```
Comment (Gouthamve) [inline discovery/file/file.go]: Make it a Gauge? @brian-brazil Do you think a counter is better or gauge?

File: discovery/file/file.go [hunk 5]
```diff
@@ -33,6 +34,48 @@ import (
...
+		)
+	}
+}
+
+//NewTimeStampManager creates a TimeStampManager
+func NewTimeStampManager() *TimeStampManager {
+	return &TimeStampManager{
+		Description: prometheus.NewDesc(
+			"prometheus_sd_file_timestamp",
+			"Timestamp of files",
```
Comment (Gouthamve) [inline discovery/file/file.go]: files read by FileSD.

File: discovery/file/file.go (hunk 4 above)
Comment (brian-brazil) [inline discovery/file/file.go]: This is a gauge, as you care about the value of it.

File: discovery/file/file.go [hunk 6]
```diff
@@ -33,6 +34,54 @@ import (
...
+	t.filenames = files
+}
+
+//Collect creates constant metrics for each file with last modified time of the file
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	files := t.filenames
+	for i := 0; i < len(files); i++ {
+		info, err := os.Stat(files[i])
+		if err != nil {
+			t.logger.Errorf("Error getting the fileinfo of the file%s: %s", files[i], err)
```
Comment (Gouthamve) [inline discovery/file/file.go]: Space b/w file and %s

Comment (grobie) [inline discovery/file/file.go]: Also best to use `%q` to log the filename, as they might contain spaces.

File: discovery/file/file.go [hunk 7]
```diff
@@ -33,6 +34,54 @@ import (
...
+			continue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			t.Description,
+			prometheus.CounterValue,
```
Comment (grobie) [inline discovery/file/file.go]: A timestamp is a gauge.

File: discovery/file/file.go [hunk 8]
```diff
@@ -33,6 +34,54 @@ import (
 
 const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
 
+//TimestampCollector is a Custom Collector for Timestamps of the files
```
Comment (grobie) [inline discovery/file/file.go]: The comment formatting style in Prometheus projects to use full sentences, always including the trailing dot and to leave a space to the comment start marker. Please update accordingly everywhere.

**Comment by KalivarapuReshma** [comment]
@grobie PTAL.

**Comment by brian-brazil** [comment]
This only handles a single file_sd_config, there could be multiple across the config file and they could be added and removed over time. You'll need to track that, and add some locking around the SetFiles.

**Pull request closed without merging**